WireDomain/Sources/WireDomain/Event Decryption/MLSMessageDecryptor.swift

```diff
@@ -0,0 +1,168 @@
+    ) async throws -> ConversationMLSMessageAddEvent
+
+}
+
+struct MLSMessageDecryptor: MLSMessageDecryptorProtocol {
+
+    let mlsDecryptionService: any MLSDecryptionServiceInterface
+    let mlsService: any MLSServiceInterface
+    let conversationLocalStore: any ConversationLocalStoreProtocol
+
+    func decryptedEventData(
+        from eventData: ConversationMLSMessageAddEvent
+    ) async throws -> ConversationMLSMessageAddEvent {
+        let conversationID = eventData.conversationID
+
+        guard let mlsConversation = await conversationLocalStore.fetchConversation(
+            id: conversationID.uuid,
+            domain: conversationID.domain
+        ) else {
+            WireLogger.mls.error(
+                "failed to add mls message: conversation not found in db"
+            )
+
+            throw MLSMessageDecryptorError.conversationNotFound
+        }
+
+        guard let mlsGroupID = await conversationLocalStore.mlsGroupID(
+            for: mlsConversation
+        ) else {
+            WireLogger.mls.error(
+                "failed to add mls message: missing MLS group ID"
+            )
+
+            throw MLSMessageDecryptorError.missingMLSGroupID
+        }
+
+        guard await conversationLocalStore.isConversationMLSReady(
+            mlsConversation
+        ) else {
+            WireLogger.mls.warn(
+                "failed to add mls message: conversation is not ready"
+            )
+
+            throw MLSMessageDecryptorError.mlsConversationNotReady
+        }
+
+        let decryptionResults = await decryptMLSMessage(
+            message: eventData.message,
+            mlsGroupID: mlsGroupID,
+            subconversation: eventData.subconversation
+        )
+
+        let decryptedMessages = await processMLSMessageDecryptionResults(
+            decryptionResults,
+            mlsConversation: mlsConversation,
+            senderID: eventData.senderID.uuid,
+            senderDomain: eventData.senderID.domain,
+            date: eventData.timestamp
+        )
+
+        var decryptedEvent = eventData
+        decryptedEvent.decryptedMessages = decryptedMessages
+
+        return decryptedEvent
+    }
+
+    private func decryptMLSMessage(
+        message: String,
+        mlsGroupID: MLSGroupID,
+        subconversation: String?
+    ) async -> [MLSDecryptResult] {
+        do {
+            let subconvType = subconversation != nil ? SubgroupType(rawValue: subconversation!) : nil
+
+            let results = try await mlsDecryptionService.decrypt(
+                message: message,
+                for: mlsGroupID,
+                subconversationType: subconvType
+            )
+
+            if results.isEmpty {
+                WireLogger.mls.info(
+                    "successfully decrypted mls message but no result was returned"
+                )
+
+                return []
+            }
+
+            return results
+
+        } catch {
+            WireLogger.mls.warn(
```

Comment:
```suggestion
            WireLogger.mls.error(
```

Reply:
done in https://github.com/wireapp/wire-ios/pull/2164/commits/b31805a437cf70da6bd6c7ebba537c3bc769a76e